File: Ashutosh-Singh/ClassTests/15th-JAN-2024/Problem1.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int firstOccurence(vector<int>& arr, int n, int k) {
        int s = 0;
        int e = n - 1;
        int ans = -1;
        while (s <= e) {
            int mid = s + (e - s) / 2;
            if (arr[mid] == k) {
                ans = mid;
                e = mid - 1; // Search on the left half
            } else if (arr[mid] < k) {
                s = mid + 1; // Search on the right half
            } else {
                e = mid - 1; // Search on the left half
            }
        }
        return ans;
    }

    int lastOccurence(vector<int>& arr, int n, int k) {
        int s = 0;
        int e = n - 1;
        int ans = -1;
        while (s <= e) {
            int mid = s + (e - s) / 2;
            if (arr[mid] == k) {
                ans = mid;
                s = mid + 1; // Search on the right half
            } else if (arr[mid] < k) {
                s = mid + 1; // Search on the right half
            } else {
                e = mid - 1; // Search on the left half
            }
        }
        return ans;
    }

    int searchdiff(vector<int>& nums, int target) {
        int n = nums.size();
        int first = firstOccurence(nums, n, target);
        if (first == -1) {
            return 0; // Target not found
        }
        int second = lastOccurence(nums, n, target);
        return second - first + 1;
    }
};

int main() {
    Solution solution;
    vector<int> nums = {1, 2, 2, 2, 3, 4, 5};
    int target = 2;

    int result = solution.searchdiff(nums, target);
    cout << "The number of occurrences of " << target << " is: " << result << endl;

    return 0;
}

File: NirbhayTyagi/BinarySearchAssessment/problem1.java
package BinarySearchAssessment;

import java.util.Scanner;
class problem1 {
    public static int numberOfOccurrences(int[] arr, int target) {
        int firstOccurence = first(arr, target);
        int lastOccurence = second(arr, target);
        return (lastOccurence - firstOccurence) < 0 ? -1: lastOccurence - firstOccurence;
    }
    private static int first(int[] arr, int target) {
        int starting = 0;
        int  ending = arr.length;
        while (starting < ending) {
            int mid = (starting + ending) / 2;
            if (arr[mid] >= target) {
                ending = mid;
            }
            else {
                starting = mid + 1;
            }
        }
        return starting;
    }
    private static int second(int[] arr, int target) {
        int starting = 0;
        int ending = arr.length;
        while (starting < ending) {
            int mid = (starting + ending) / 2;
            if (arr[mid] <= target) {
                starting = mid + 1;
            } else{
                ending = mid;
            }
        }
        return starting;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter size of array: ");
        int size = sc.nextInt();
        int[] arr = new int[size];
        System.out.println("Enter elements in a sorted order:");
        for (int i = 0; i < size; i++) {
            System.out.println("Enter element: ");
            arr[i] = sc.nextInt();
        }
        System.out.println("Enter target: ");
        int target = sc.nextInt();
        System.out.println("Number of " + target + ": " + numberOfOccurrences(arr, target));
        sc.close();
    }
}

//Complexity of whole program:
//Time: O(n) + O(log n)
//Space: O(n)

//Complexity of function:
//Time: O(log n)
//Space: O(1)
